C/STrace/EtwLogger.h

```diff
@@ -0,0 +1,314 @@
+	if (eventMetadata == NULL)
+	{
+		return EVENT_DATA_DESCRIPTOR{};
+	}
+
+	RtlSecureZeroMemory(eventMetadata, eventMetadataLength);
+
+	// Set the first three fields, metadata about the event.
+	eventMetadata->TotalLength = eventMetadataLength;
+	eventMetadata->Tag = 0;
+	strcpy(eventMetadata->EventName, eventName);
+
+	// Set the metadata for each field.
+	uint8_t* currentLocation = ((uint8_t*)eventMetadata) + eventMetadataHeaderLength;
+	detail::SetFieldMetadata(currentLocation, args...);
+
+	// Create an EVENT_DATA_DESCRIPTOR pointing to the metadata.
+	EVENT_DATA_DESCRIPTOR eventMetadataDesc;
+	EventDataDescCreate(&eventMetadataDesc, eventMetadata, eventMetadata->TotalLength);
+	eventMetadataDesc.Type = EVENT_DATA_DESCRIPTOR_TYPE_EVENT_METADATA;  // Descriptor contains event metadata.
+
+	return eventMetadataDesc;
+}
+
+// final recursion case, string-specific overload
+NTSTATUS CreateTracePropertyRecursive(OUT PEVENT_DATA_DESCRIPTOR propertyDataDescriptors, const char* fieldName, int fieldType, const char* fieldValue)
+{
+	// fieldName and fieldType are used in the event metadata descriptor, not here.
+	UNREFERENCED_PARAMETER(fieldName);
+	UNREFERENCED_PARAMETER(fieldType);
+
+	// Copy the input value to its own space.
+	const auto newSpace = (char*)ExAllocatePoolWithTag(NonPagedPoolNx, strlen(fieldValue) + 1, 'wteE');
+	if (newSpace == NULL)
+	{
+		return STATUS_UNSUCCESSFUL;
+	}
+	strcpy(newSpace, fieldValue);
+
+	// Create the event data descriptor pointing to the value of the field.
+	EventDataDescCreate(OUT propertyDataDescriptors, newSpace, (ULONG)(strlen(fieldValue) + 1));
+
+	return STATUS_SUCCESS;
+}
+
+// final recursion case
+template<typename FieldValue>
+NTSTATUS CreateTracePropertyRecursive(OUT PEVENT_DATA_DESCRIPTOR propertyDataDescriptors, const char* fieldName, int fieldType, FieldValue& fieldValue)
+{
+	// fieldName and fieldType are used in the event metadata descriptor, not here.
+	UNREFERENCED_PARAMETER(fieldName);
+	UNREFERENCED_PARAMETER(fieldType);
+
+	// Copy the input value to its own space.
+	const auto newSpace = ExAllocatePoolWithTag(NonPagedPoolNx, sizeof(FieldValue), 'wteE');
+	if (newSpace == NULL)
+	{
+		return STATUS_UNSUCCESSFUL;
+	}
+	memcpy(newSpace, &fieldValue, sizeof(FieldValue));
+
+	// Create the event data descriptor pointing to the value of the field.
+	EventDataDescCreate(OUT propertyDataDescriptors, newSpace, sizeof(FieldValue));
+
+	return STATUS_SUCCESS;
+}
+
+// string-specific overload
+template<typename... Rest>
+NTSTATUS CreateTracePropertyRecursive(OUT PEVENT_DATA_DESCRIPTOR propertyDataDescriptors, const char* fieldName, int fieldType, const char* fieldValue, Rest... rest)
+{
+	// fieldName and fieldType are used in the event metadata descriptor, not here.
+	UNREFERENCED_PARAMETER(fieldName);
+	UNREFERENCED_PARAMETER(fieldType);
+
+	// Copy the input value to its own space.
+	const auto newSpace = (char*)ExAllocatePoolWithTag(NonPagedPoolNx, strlen(fieldValue) + 1, 'wteE');
```

Comment:
Is this pool tag a standard microsoft one, do we need to worry about conflicting with a tag that already exists?

Reply:
Ooh I've no idea to be honest! Would you know if I can find out?

Reply:
You can use poolmon or windbg to see pooltags, but I was more wondering if you got these tags from reversing dtrace. If they're just made up by you then I'd imagine they're fine.

Reply:
A 'conflict' would not be a big deal, it's just good for us to use our own for debug

Reply:
Never responded to this, but yes I made the tags myself